gtsam/hybrid/DCMixtureFactor.h

```diff
@@ -34,56 +33,79 @@ namespace gtsam {
  */
 template <class NonlinearFactorType>
 class DCMixtureFactor : public DCFactor {
- private:
-  DiscreteKey dk_;
-  std::vector<NonlinearFactorType> factors_;
-  bool normalized_;
-
  public:
   using Base = DCFactor;
   using shared_ptr = boost::shared_ptr<DCMixtureFactor>;
 
+  using FactorDecisionTree =
+      DecisionTree<Key, boost::shared_ptr<NonlinearFactorType>>;
+
+ private:
+  FactorDecisionTree factors_;
+  bool normalized_;
+
+ public:
   DCMixtureFactor() = default;
 
-  DCMixtureFactor(const KeyVector& keys, const DiscreteKey& dk,
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
+                  const FactorDecisionTree& factors, bool normalized = false)
+      : Base(keys, discreteKeys), factors_(factors), normalized_(normalized) {}
+
+  /**
+   * @brief Convenience constructor that generates the underlying factor
+   * decision tree for us.
+   *
+   * Here it is important that the vector of discrete keys and the vector of
+   * factors have a 1-to-1 mapping so that the decision tree is constructed
+   * accordingly.
+   *
+   * @param keys Vector of keys for continuous factors.
+   * @param discreteKeys Vector of discrete keys.
+   * @param factors Vector of factors. Each factor should map to its
+   * corresponding assigned discrete key.
+   * @param normalized Flag indicating if the factor error is already
+   * normalized.
+   */
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
                   const std::vector<NonlinearFactorType>& factors,
                   bool normalized = false)
-      : dk_(dk), factors_(factors), normalized_(normalized) {
-    // Compiler doesn't like `keys_` in the initializer list.
-    keys_ = keys;
-
-    // Add `dk` to `dkeys` list.
-    discreteKeys_.push_back(dk);
+      : Base(keys, discreteKeys), normalized_(normalized) {
+    std::vector<boost::shared_ptr<NonlinearFactorType>> factor_pointers;
+    for (auto&& factor : factors) {
+      boost::shared_ptr<NonlinearFactorType> f_ptr =
+          boost::make_shared<NonlinearFactorType>(factor);
+      factor_pointers.push_back(f_ptr);
+    }
+    // Generate decision tree based on discreteKey to factor mapping.
+    factors_ = FactorDecisionTree(discreteKeys, factor_pointers);
   }
 
-  /// Discrete key selecting mixture component
-  const DiscreteKey& discreteKey() const { return dk_; }
-
   DCMixtureFactor& operator=(const DCMixtureFactor& rhs) {
     Base::operator=(rhs);
-    this->dk_ = rhs.dk_;
     this->factors_ = rhs.factors_;
   }
 
   ~DCMixtureFactor() = default;
 
   double error(const Values& continuousVals,
                const DiscreteValues& discreteVals) const override {
-    // Retrieve the assignment to our discrete key.
-    const size_t assignment = discreteVals.at(dk_.first);
+    /********************************************************/
```

Comment:
Huh?

Reply:
Oh this was made to demarcate the old version vs the new version so Fan and I knew what were we doing. Let me remove it.